From 21b411220215cfe2ab3892d05b9dfb61bee7c38c Mon Sep 17 00:00:00 2001
From: Sam Brightman <sam.brightman@gmail.com>
Date: Sun, 23 Oct 2022 09:52:08 +0100
Subject: [PATCH] Squash of bpo-42504 fixes, backported to 2.7

[f9fddfd29ac11b7224569ec3da6871bb529c76ab]
Squash of bpo-42504 fixes, backported to 3.7

Co-authored-by: Bo Anderson <mail@boanderson.me>

[5291639e611dc3f55a34666036f2c3424648ba50]
bpo-42504: fix for MACOSX_DEPLOYMENT_TARGET=11 (GH-23556)

macOS releases numbering has changed as of macOS 11 Big Sur.  Previously, major releases were of the form 10.x, 10.x+1, 10.x+2, etc; as of Big Sur, they are now x, x+1, etc, so, for example, 10.15, 10.15.1, ..., 10.15.7, 11, 11.0.1, 11.1, ..., 12, 12.1, etc. Allow Python to build with single-digit deployment target values. Patch provided by FX Coudert.

Co-authored-by: FX Coudert <fxcoudert@gmail.com>
Co-authored-by: Mitchell Hentges <mhentges@mozilla.com> (backported to 3.7)

[49926cf2bcc8b44d9b8f148d81979ada191dd9d5]
bpo-42504: Ensure that get_config_var(MACOSX_DEPLOYMENT_TARGET) is a string (GH-24341)

Co-authored-by: Ronald Oussoren <ronaldoussoren@mac.com>
---
 Lib/distutils/tests/test_build_ext.py |  6 +++++-
 Lib/sysconfig.py                      | 21 ++++++++++++++++++---
 2 files changed, 23 insertions(+), 4 deletions(-)

diff --git a/Lib/distutils/tests/test_build_ext.py b/Lib/distutils/tests/test_build_ext.py
index a6d2d2eb15..dd6b3bbfc8 100644
--- a/Lib/distutils/tests/test_build_ext.py
+++ b/Lib/distutils/tests/test_build_ext.py
@@ -496,7 +496,11 @@ class BuildExtTestCase(support.TempdirManager,
             target = '%02d%01d0' % target
         else:
             # for 10.10 and beyond -> "10nn00"
-            target = '%02d%02d00' % target
+            if len(target) >= 2:
+                target = '%02d%02d00' % target
+            else:
+                # 11 and later can have no minor version (11 instead of 11.0)
+                target = '%02d0000' % target
         deptarget_ext = Extension(
             'deptarget',
             [deptarget_c],
diff --git a/Lib/sysconfig.py b/Lib/sysconfig.py
index 9c8350d953..9da7bfd4e8 100644
--- a/Lib/sysconfig.py
+++ b/Lib/sysconfig.py
@@ -5,6 +5,11 @@ import sys
 import os
 from os.path import pardir, realpath
 
+# Keys for get_config_var() that are never converted to Python integers.
+_ALWAYS_STR = {
+    'MACOSX_DEPLOYMENT_TARGET',
+}
+
 _INSTALL_SCHEMES = {
     'posix_prefix': {
         'stdlib': '{base}/lib/python{py_version_short}',
@@ -228,6 +233,9 @@ def _parse_makefile(filename, vars=None):
                 notdone[n] = v
             else:
                 try:
+                    if n in _ALWAYS_STR:
+                        raise ValueError
+
                     v = int(v)
                 except ValueError:
                     # insert literal `$'
@@ -259,7 +267,10 @@ def _parse_makefile(filename, vars=None):
                     if "$" in after:
                         notdone[name] = value
                     else:
-                        try: value = int(value)
+                        try:
+                            if n in _ALWAYS_STR:
+                                raise ValueError
+                            value = int(value)
                         except ValueError:
                             done[name] = value.strip()
                         else:
@@ -397,8 +408,12 @@ def parse_config_h(fp, vars=None):
         m = define_rx.match(line)
         if m:
             n, v = m.group(1, 2)
-            try: v = int(v)
-            except ValueError: pass
+            try:
+                if n in _ALWAYS_STR:
+                    raise ValueError
+                v = int(v)
+            except ValueError:
+                pass
             vars[n] = v
         else:
             m = undef_rx.match(line)
-- 
2.38.1

